File: tests/functional/Bots/JavaScript/Consumers/CodeFirst/WaterfallHostBot/dialogs/tangentDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { InputHints, MessageFactory } = require('botbuilder');
const { ComponentDialog, DialogSet, TextPrompt, WaterfallDialog, DialogTurnStatus } = require('botbuilder-dialogs');

const TANGENT_DIALOG = 'TangentDialog';
const WATERFALL_DIALOG = 'WaterfallDialog';
const TEXT_PROMPT = 'TextPrompt';

class TangentDialog extends ComponentDialog {
  /**
   * @param {string} dialogId
   */
  constructor (dialogId = TANGENT_DIALOG) {
    super(dialogId);

    this.addDialog(new TextPrompt(TEXT_PROMPT));
    this.addDialog(new WaterfallDialog(WATERFALL_DIALOG, [
      this.step1.bind(this),
      this.step2.bind(this),
      this.endStep.bind(this)
    ]));

    this.initialDialogId = WATERFALL_DIALOG;
  }

  /**
   * The run method handles the incoming activity (in the form of a TurnContext) and passes it through the dialog system.
   * If no dialog is active, it will start the default dialog.
   * @param {import('botbuilder').TurnContext} turnContext
   * @param {*} accessor
   */
  async run (turnContext, accessor) {
    const dialogSet = new DialogSet(accessor);
    dialogSet.add(this);

    const dialogContext = await dialogSet.createContext(turnContext);
    const results = await dialogContext.continueDialog();
    if (results.status === DialogTurnStatus.empty) {
      await dialogContext.beginDialog(this.id);
    }
  }

  /**
   * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
   */
  async step1 (stepContext) {
    const messageText = 'Tangent step 1 of 2, say something.';
    const promptMessage = MessageFactory.text(messageText, messageText, InputHints.ExpectingInput);

    return stepContext.prompt(TEXT_PROMPT, promptMessage);
  }

  /**
   * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
   */
  async step2 (stepContext) {
    const messageText = 'Tangent step 2 of 2, say something.';
    const promptMessage = MessageFactory.text(messageText, messageText, InputHints.ExpectingInput);

    return stepContext.prompt(TEXT_PROMPT, promptMessage);
  }

  /**
   * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
   */
  async endStep (stepContext) {
    return stepContext.endDialog();
  }
}

module.exports.TangentDialog = TangentDialog;

File: tests/functional/Bots/Python/Consumers/CodeFirst/WaterfallHostBot/dialogs/tangent_dialog.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.dialogs import (
    ComponentDialog,
    WaterfallDialog,
    WaterfallStepContext,
    TextPrompt,
    PromptOptions,
)
from botbuilder.core import MessageFactory
from botbuilder.schema import InputHints


class TangentDialog(ComponentDialog):
    """
    A simple waterfall dialog used to test triggering tangents from "MainDialog".
    """

    def __init__(self):
        super().__init__(TangentDialog.__name__)

        self.add_dialog(TextPrompt(TextPrompt.__name__))
        self.add_dialog(
            WaterfallDialog(
                WaterfallDialog.__name__, [self._step_1, self._step_2, self._end_step]
            )
        )

        self.initial_dialog_id = WaterfallDialog.__name__

    async def _step_1(self, step_context: WaterfallStepContext):
        message_text = "Tangent step 1 of 2, say something."
        prompt_message = MessageFactory.text(
            message_text, message_text, InputHints.expecting_input
        )
        return await step_context.prompt(
            TextPrompt.__name__, PromptOptions(prompt=prompt_message)
        )

    async def _step_2(self, step_context: WaterfallStepContext):
        message_text = "Tangent step 2 of 2, say something."
        prompt_message = MessageFactory.text(
            message_text, message_text, InputHints.expecting_input
        )
        return await step_context.prompt(
            TextPrompt.__name__, PromptOptions(prompt=prompt_message)
        )

    async def _end_step(self, step_context: WaterfallStepContext):
        return await step_context.end_dialog()
